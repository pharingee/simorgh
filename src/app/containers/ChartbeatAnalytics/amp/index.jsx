import React from 'react';
import { string, number } from 'prop-types';

const chartbeatAmpConfigOptions = options => ({
  vars: {
    uid: options.chartbeatUID,
    sections: options.sections,
    domain: options.domain,
    contentType: options.type,
    idSync: {
      bbc_hid: options.cookie,
    },
  },
});

const JsonInlinedScript = data => (
  <script
    type="application/json"
    /* eslint-disable-next-line react/no-danger */
    dangerouslySetInnerHTML={{ __html: JSON.stringify(data) }}
  />
);

const AmpChartbeatBeacon = ({
  domain,
  type,
  sections,
  cookie,
  chartbeatUID,
}) => (
  <amp-analytics type="chartbeat">
    {JsonInlinedScript(
      chartbeatAmpConfigOptions({
        domain,
        type,
        sections,
        cookie,
        chartbeatUID,
      }),
    )}
  </amp-analytics>
);

AmpChartbeatBeacon.propTypes = {
  domain: string.isRequired,
  type: string.isRequired,
  sections: string.isRequired,
  cookie: string,
  chartbeatUID: number.isRequired,
};

AmpChartbeatBeacon.defaultProps = {
  cookie: null,
};

export default AmpChartbeatBeacon;
