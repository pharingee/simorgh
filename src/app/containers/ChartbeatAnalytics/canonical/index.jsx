import React from 'react';
import { string, bool, number } from 'prop-types';
import Helmet from 'react-helmet';

const CanonicalChartbeatBeacon = ({
  domain,
  type,
  sections,
  cookie,
  chartbeatUID,
  useCanonical,
  chartbeatSource,
}) => (
  <Helmet>
    <script async type="text/javascript">
      {`
        (function(){
          var _sf_async_config = window._sf_async_config = (window._sf_async_config || {});
          _sf_async_config.uid = ${chartbeatUID};
          _sf_async_config.domain = "${domain}";
          _sf_async_config.useCanonical = ${useCanonical};
          _sf_async_config.useCanonicalDomain = ${useCanonical};
          _sf_async_config.contentType = "${type}";
          _sf_async_config.sections = "${sections}";
          _sf_async_config.idSync = {
           bbc_hid: "${cookie}"
          };
          function loadChartbeat() {
           var e = document.createElement('script');
           var n = document.getElementsByTagName('script')[0];
           e.type = 'text/javascript';
           e.async = true;
           e.src = "${chartbeatSource}";
           n.parentNode.insertBefore(e, n);
          }
          loadChartbeat();
        })();
      `}
    </script>
  </Helmet>
);

CanonicalChartbeatBeacon.propTypes = {
  domain: string.isRequired,
  type: string.isRequired,
  sections: string.isRequired,
  cookie: string,
  chartbeatUID: number.isRequired,
  useCanonical: bool.isRequired,
  chartbeatSource: string.isRequired,
};

CanonicalChartbeatBeacon.defaultProps = {
  cookie: null,
};

export default CanonicalChartbeatBeacon;
