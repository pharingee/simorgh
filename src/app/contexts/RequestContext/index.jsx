import React from 'react';
import { bool, node, oneOf, string } from 'prop-types';
import getStatsDestination from './getStatsDestination';
import getStatsPageIdentifier from './getStatsPageIdentifier';
import getOriginContext from './getOriginContext';
import getEnv from './getEnv';

export const RequestContext = React.createContext('default');

export const RequestContextProvider = ({
  children,
  bbcOrigin,
  id,
  isAmp,
  pageType,
  service,
  previousPath,
}) => {
  const { isUK, origin } = getOriginContext(bbcOrigin);
  const env = getEnv(origin);
  const platform = isAmp ? 'amp' : 'canonical';
  const statsDestination = getStatsDestination({
    isUK,
    env,
    service,
  });
  const statsPageIdentifier = getStatsPageIdentifier({
    pageType,
    service,
    id,
  });

  const value = {
    env,
    id,
    isUK,
    origin,
    pageType,
    platform,
    statsDestination,
    statsPageIdentifier,
    previousPath,
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  bbcOrigin: string,
  children: node.isRequired,
  id: string,
  isAmp: bool.isRequired,
  pageType: oneOf(['article', 'frontPage', 'media']).isRequired,
  service: string.isRequired,
  previousPath: string,
};

RequestContextProvider.defaultProps = {
  bbcOrigin: null,
  id: null,
  previousPath: null,
};
