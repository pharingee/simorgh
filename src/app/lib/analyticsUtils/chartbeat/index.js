import Cookie from 'js-cookie';
import onClient from '../../utilities/onClient';

const ID_COOKIE = 'ckns_sylphid';

export const chartbeatUID = 50924;
export const useCanonical = true;
export const chartbeatSource = '//static.chartbeat.com/js/chartbeat.js';

const buildSectionArr = (service, value, type) => [
  `${service} - ${value}`,
  ...(type ? [`${service} - ${value} - ${type}`] : []),
];

const buildServiceType = (service, type) => [`${service} - ${type}`];

const capitalize = s => s.charAt(0).toUpperCase() + s.slice(1);

export const getSylphidCookie = () =>
  onClient() ? Cookie.get(ID_COOKIE) : null;

export const getDomain = service => {
  const serviceLower = service.toLowerCase();

  return serviceLower === 'news' ? 'bbc.co.uk' : `${service}.bbc.co.uk`;
};

export const getType = (pageType, shorthand = false) => {
  switch (pageType) {
    case 'frontPage':
    case 'index':
      return shorthand ? 'IDX' : 'Index';
    case 'article':
      return shorthand ? 'ART' : 'New Article';
    default:
      return null;
  }
};

export const buildSections = (service, type, producer, chapter) => {
  const addProducer = producer && service !== producer;
  const serviceCap = capitalize(service);
  const pageType = getType(type, true);

  const parts = [
    serviceCap,
    ...(type ? buildServiceType(serviceCap, pageType) : []),
    ...(addProducer ? buildSectionArr(serviceCap, producer, pageType) : []),
    ...(chapter ? buildSectionArr(serviceCap, chapter, pageType) : []),
  ];

  return parts.join(', ');
};
