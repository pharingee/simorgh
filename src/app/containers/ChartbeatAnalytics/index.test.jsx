import React from 'react';
import { node, string } from 'prop-types';
import renderer from 'react-test-renderer';
import { RequestContextProvider } from '../../contexts/RequestContext';
import { ServiceContextProvider } from '../../contexts/ServiceContext';
import { ToggleContextProvider } from '../../contexts/ToggleContext';
import ChartBeatAnalytics from '.';
import * as testUtils from '../../lib/analyticsUtils/chartbeat';
import * as amp from './amp';
import * as canonical from './canonical';

const ContextWrap = ({ pageType, platform, origin, children }) => (
  <RequestContextProvider
    isAmp={platform === 'amp'}
    pageType={pageType}
    service="news"
    bbcOrigin={origin}
  >
    <ServiceContextProvider service="news">
      <ToggleContextProvider>{children}</ToggleContextProvider>
    </ServiceContextProvider>
  </RequestContextProvider>
);

ContextWrap.propTypes = {
  children: node.isRequired,
  pageType: string.isRequired,
  origin: string.isRequired,
  platform: string.isRequired,
};

describe('Charbeats Analytics Container', () => {
  it('should call CanonicalCharbeatsBeacon when platform is canonical, and toggle enabled for chartbeat for local', () => {
    const mockCanonical = jest.fn().mockReturnValue('canonical-return-value');
    const { chartbeatSource } = testUtils;
    canonical.default = mockCanonical;
    testUtils.getDomain = jest
      .fn()
      .mockImplementation(service => `${service}-domain`);

    testUtils.getSylphidCookie = jest.fn().mockImplementation(() => 'cookie');

    testUtils.getType = jest.fn().mockImplementation(() => 'article');
    testUtils.buildSections = jest
      .fn()
      .mockImplementation(() => 'secction1 section2');

    const tree = renderer
      .create(
        <ContextWrap
          platform="canonical"
          pageType="article"
          origin="localhost.bbc.com"
        >
          <ChartBeatAnalytics />
        </ContextWrap>,
      )
      .toJSON();

    expect(mockCanonical).toHaveBeenCalledTimes(1);
    expect(mockCanonical).toHaveBeenCalledWith(
      {
        chartbeatUID: 50924,
        cookie: 'cookie',
        domain: 'test-domain',
        sections: 'secction1 section2',
        type: 'article',
        useCanonical: true,
        chartbeatSource,
      },
      {},
    );
    expect(testUtils.getDomain).toHaveBeenCalledTimes(1);
    expect(testUtils.getSylphidCookie).toHaveBeenCalledTimes(1);
    expect(testUtils.getType).toHaveBeenCalledTimes(1);
    expect(testUtils.buildSections).toHaveBeenCalledTimes(1);
    expect(tree).toMatchSnapshot();
  });
  it('should call AmpCharbeatsBeacon when platform is amp and toggle enabled for chartbeat on test', () => {
    const mockAmp = jest.fn().mockReturnValue('amp-return-value');
    amp.default = mockAmp;

    testUtils.getDomain = jest
      .fn()
      .mockImplementation(service => `${service}-domain`);

    testUtils.getSylphidCookie = jest.fn().mockImplementation(() => 'cookie');

    testUtils.getType = jest.fn().mockImplementation(() => 'article');
    testUtils.buildSections = jest
      .fn()
      .mockImplementation(() => 'secction1 section2');

    const tree = renderer
      .create(
        <ContextWrap platform="amp" pageType="article" origin="test.bbc.com">
          <ChartBeatAnalytics />
        </ContextWrap>,
      )
      .toJSON();
    expect(mockAmp).toHaveBeenCalledTimes(1);
    expect(mockAmp).toHaveBeenCalledWith(
      {
        chartbeatUID: 50924,
        cookie: 'cookie',
        domain: 'test-domain',
        sections: 'secction1 section2',
        type: 'article',
      },
      {},
    );
    expect(testUtils.getDomain).toHaveBeenCalledTimes(1);
    expect(testUtils.getSylphidCookie).toHaveBeenCalledTimes(1);
    expect(testUtils.getType).toHaveBeenCalledTimes(1);
    expect(testUtils.buildSections).toHaveBeenCalledTimes(1);
    expect(tree).toMatchSnapshot();
  });
  it('should return null when toggle is disbaled for live', () => {
    const tree = renderer
      .create(
        <ContextWrap platform="canonical" pageType="article" origin="bbc.com">
          <ChartBeatAnalytics />
        </ContextWrap>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
