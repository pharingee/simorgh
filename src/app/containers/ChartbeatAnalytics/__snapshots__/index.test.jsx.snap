// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charbeats Analytics Container should call AmpCharbeatsBeacon when platform is amp and toggle enabled for chartbeat on test 1`] = `"amp-return-value"`;

exports[`Charbeats Analytics Container should call CanonicalCharbeatsBeacon when platform is canonical, and toggle enabled for chartbeat for local 1`] = `"canonical-return-value"`;

exports[`Charbeats Analytics Container should return null when toggle is disbaled for live 1`] = `null`;
